<!--<img src='./figures/logo-ecole-polytechnique-ve.jpg' style='position:absolute; top:0; right:0;' width='100px' height='' alt='' />-->

<center>**Bachelor of Ecole Polytechnique**</center>
<center>Computational Mathematics, year 2, semester 1</center>
<center>Lecturer: Lucas Gerin <a href="mailto:lucas.gerin@polytechnique.edu">(send mail)</a></center>

# Symbolic computation 3: Test


## Table of contents

- [Exercise 1: Rational fractions](#continued)
- [Exercise 2. A polynomial riddle](#exo_web) 
- [Exercise 3: Multivariate polynomial](#multivariate)
- [Exercise 4: A linear recurrence](#linear_recurrence)
- [Exercise 5: Square roots](#square)


In [1]:
# execute this part to modify the css style
from IPython.core.display import HTML
def css_styling():
    styles = open("./style/custom2.css").read()
    return HTML(styles)
css_styling()


In [4]:
## loading python libraries

# necessary to display plots inline:
%matplotlib inline   

# load the libraries
import matplotlib.pyplot as plt # 2D plotting library
import numpy as np              # package for scientific computing  
from pylab import *

from math import *              # package for mathematics (pi, arctan, sqrt, factorial ...)
import sympy as sympy             # package for symbolic computation
from sympy import *


# Please read!

<br><br>

<div markdown=1 class=Abstract>

### Guidelines for today

* You will use SymPy to solve the above exercises. The whole Notebook can be done with the following SymPy functions:
```
expand(),factor(),Rational(),Matrix(),simplify(),solve()
```
* In order to check your formulas do not hesitate to use `N()` to obtain a numerical evaluation.
* Remember that in SymPy you have to use `Rational` in order to prevent numerical evaluation of fractions. When you use symbolic variables, do not use `Rational`. For example: write `a/b` instead of `Rational(a,b)`.

## <a id='continued'></a>
### Exercise 1. Rational fractions.

<div markdown=1 class="DoIt"> 

We set
$$
u_1=\frac{1}{3+1}, \quad u_2=\frac{1}{3+\frac{1}{3+1}},\quad u_3=\frac{1}{3+\frac{1}{3+\frac{1}{3+1}}}, \quad u_4=\frac{1}{3+\frac{1}{3+\frac{1}{3+\frac{1}{3+1}}}}, \dots
$$
Using SymPy, write $u_{25}$ as a rational fraction $a/b$.

<i>(Hint: This one is easier than the fraction in last Notebook. Here you can find a simple relation between $u_n$ and $u_{n+1}$.)</i>

In [6]:
# Initialization
u1 = Rational(1, 4)
u2 = Rational(1, 3 + Rational(1, 4))


def un(n):
    if n == 1:
        return u1
    # Recursive definition of the sequence
    return Rational(1, 3 + un(n - 1))


for n in range(1, 5):
    print('u_{} = {}'.format(n, un(n)))

print('-------------------------')
print('Answer 1')
print('u_25 = {}'.format(un(25)))


u_1 = 1/4
u_2 = 4/13
u_3 = 13/43
u_4 = 43/142
-------------------------
Answer 1
u_25 = 3387464586001/11188035508324


<div markdown=1 class="Answers"> 

**Answer 1**

It is easy to see that $u_{n+1}=\frac{1}{3+u_n}$ from the sequence. Therefore, we can simply use the recursion formula to obtain $u_{25}$. By computation we got
$$
u_{25}=\frac{3387464586001}{11188035508324}
$$


## <a id='exo_web'></a>
### Exercise 2. A polynomial riddle

<div markdown=1 class="DoIt"> On the web one can find the following riddle:
<center>
If $x+\frac{1}{x}=5$, what is the value of $x^7+\left(\frac{1}{x}\right)^7$ ?
</center>
Solve this with SymPy.

In [19]:
# Set up variables
var('x')
# use sympy to solve the equation x+1/x=5
print('-------------------------')
print('Answer 2')
solution = solve(x + 1 / x - 5, x)
print('The first solution is')
display(solution[0])
print('The Second solution is')
display(solution[1])


def f(x):
    return x**7 + (1 / x)**7

display(simplify(f(solution[0])))
display(simplify(f(solution[1])))

print('-------------------------')
print('The value of x^7+1/x^7 is ', simplify(f(solution[0])))

-------------------------
Answer 2
The first solution is


5/2 - sqrt(21)/2

The Second solution is


sqrt(21)/2 + 5/2

57965

57965

-------------------------
The value of x^7+1/x^7 is  57965


<div markdown=1 class="Answers"> 

**Answer for Exercise 2**

We can use the `solve` function to solve the equation $x+\frac{1}{x}=5$. We get two solutions which are $\frac{5}{2}-\frac{\sqrt{21}}{2}$ and $\frac{5}{2}+\frac{\sqrt{21}}{2}$. Then we can use the function `f(x)` we defined to compute the result. For both solutions we have the same result of $x^7+\left(\frac{1}{x}\right)^7=57965$.

## <a id='multivariate'></a>
### Exercise 3. Multivariate Polynomials 

<div markdown=1 class="DoIt"> 

Use SymPy to find the coefficient of $x^3y^2z^7$ in the polynomial $(x+2y+z)^{12}$.

In [24]:
var('x y z')
poly = (x + 2*y + z) ** 12
poly_expand = expand(poly)
display(poly_expand)

# And we try to find the coefficients x^3 y^2 z^7
print('-------------------------')
print('Answer 3')
print('The coefficient of x^3 y^2 z^7 is ', poly_expand.coeff(x**3 * y**2 * z**7))

x**12 + 24*x**11*y + 12*x**11*z + 264*x**10*y**2 + 264*x**10*y*z + 66*x**10*z**2 + 1760*x**9*y**3 + 2640*x**9*y**2*z + 1320*x**9*y*z**2 + 220*x**9*z**3 + 7920*x**8*y**4 + 15840*x**8*y**3*z + 11880*x**8*y**2*z**2 + 3960*x**8*y*z**3 + 495*x**8*z**4 + 25344*x**7*y**5 + 63360*x**7*y**4*z + 63360*x**7*y**3*z**2 + 31680*x**7*y**2*z**3 + 7920*x**7*y*z**4 + 792*x**7*z**5 + 59136*x**6*y**6 + 177408*x**6*y**5*z + 221760*x**6*y**4*z**2 + 147840*x**6*y**3*z**3 + 55440*x**6*y**2*z**4 + 11088*x**6*y*z**5 + 924*x**6*z**6 + 101376*x**5*y**7 + 354816*x**5*y**6*z + 532224*x**5*y**5*z**2 + 443520*x**5*y**4*z**3 + 221760*x**5*y**3*z**4 + 66528*x**5*y**2*z**5 + 11088*x**5*y*z**6 + 792*x**5*z**7 + 126720*x**4*y**8 + 506880*x**4*y**7*z + 887040*x**4*y**6*z**2 + 887040*x**4*y**5*z**3 + 554400*x**4*y**4*z**4 + 221760*x**4*y**3*z**5 + 55440*x**4*y**2*z**6 + 7920*x**4*y*z**7 + 495*x**4*z**8 + 112640*x**3*y**9 + 506880*x**3*y**8*z + 1013760*x**3*y**7*z**2 + 1182720*x**3*y**6*z**3 + 887040*x**3*y**5*z**4 + 443520*

-------------------------
Answer 3
The coefficient of x^3 y^2 z^7 is  31680


<div markdown=1 class="Answers"> 

**Answer for Exercise 3**

We can use the `expand` function to expand the polynomial $(x+2y+z)^{12}$. Then we can use the `coeff` function to find the coefficient of $x^3y^2z^7$ is 31680.

## <a id="linear_recurrence"></a>
### Exercise 4. A linear recurrence

<div markdown=1 class="DoIt"> 

Let $(u_n)$ be defined by
\begin{equation}
\begin{cases}
u_0&=7,\\
u_1&=12,\\
u_{n}&=u_{n-2} +8n-2. \qquad (\forall n\geq 2).
\end{cases}
\tag{$\star$}
\end{equation}
1. Let  $a,b,c$ be four real parameters. Let $(w_n)_{n\geq 0}$ be the sequence defined by
$$
w_n=an^2 +bn+c.
$$
Use `SymPy` to find $a,b,c$ such that for every $0\leq n\leq 2$,
$$
w_n=u_n.
$$
2. Prove with `SymPy` that $(w_n)$ defined as above is the unique solution of ($\star$).



In [31]:
# ------------- Question 1 --------------
u_0 = 7
u_1 = 12


def un(n):
    if n == 0:
        return u_0
    if n == 1:
        return u_1
    # Recursive definition of the sequence
    return un(n - 2) + 8 * n - 2


print('u_2 = {}'.format(un(2)))
var('n')
a, b, c = sympy.symbols('a b c')
eq0 = Eq(c, 7)
eq1 = Eq(a + b + c, 12)
eq2 = Eq(4 * a + 2 * b + c, 21)
result = sympy.solve([eq1, eq2, eq0], (a, b, c))
print('-------------------------')
print('Answer')
print(result)


u_2 = 21
-------------------------
Answer
{a: 2, b: 3, c: 7}


<div markdown=1 class="Answers"> 

**Answer for Exercise 4.1**

According to the function above we have
$$
a = 2, b = 3, c=7
$$
With this choice of parameters $(w_n)$ and $(u_n)$ are same with $0\leq n \leq2$

In [43]:
#-------------- Question 2 ----------------
print('-------------------------')
a1 = result[a]
b1 = result[b]
c1 = result[c]

# we define the function w here
def w(a, b, c, n):
    return a * n**2 + b * n + c

expr = w(a1, b1, c1, n) - w(a1, b1, c1, n - 2), - 8 * n + 2
display(simplify(expr))

print('-------------------------')
print('Safety check:')
print('First values of w_n: '+str([w(a1, b1, c1, i) for i in range(10)]))

# In order to check our results:
def Recurrence(n):
    if n == 0:
        return 7
    if n == 1:
        return 12
    return Recurrence(n - 2) + 8 * n - 2

print('First values (with recurrence formula) : '+str([Recurrence(n) for n in range(10)]))


-------------------------


(2*n**2 - 2*(n - 2)**2 + 6, 2 - 8*n)

-------------------------
Safety check:
First values of w_n: [7, 12, 21, 34, 51, 72, 97, 126, 159, 196]
First values (with recurrence formula) : [7, 12, 21, 34, 51, 72, 97, 126, 159, 196]


<div markdown=1 class="Answers"> 

**Answer for Exercise 4.2**

We first observe that eq.$(\star)$ has a unique solution since the sequence $(u_n)$ is a recurrence of order one: it is uniquely determined by $u_0$.
Besides, set
$$
w_n=2n^{2}+3n+7.
$$
Parameters are chosen so that $w_0=u_0=7$.
The above script shows that for every $n$ we have that $w_n-2w_{n-2}-8n+2=0$. Therefore $(w_n)$ satisfies the recurrence $(\star)$ and starts with the same initial value $w_0=7$. Therefore $w_n=u_n$ for every $n$.

## <a id="square"></a>
### Exercise 5. Square roots

<div markdown=1 class="DoIt"> <b>Theory</b>

1. We admit that for every $n\geq 1$ there exists three integers $a_n,b_n,c_n,d_n$ such that 
$$
(\sqrt{2}+\sqrt{3})^n=a_n+b_n\sqrt{2}+c_n\sqrt{3}+d_n\sqrt{6}.
$$
Find a $4\times 4$ matrix $A$ such that
$$
\begin{pmatrix}
a_{n+1}\\
b_{n+1}\\
c_{n+1}\\
d_{n+1}
\end{pmatrix}
=
A\times 
\begin{pmatrix}
a_{n}\\
b_{n}\\
c_{n}\\
d_{n}
\end{pmatrix}
$$
2. Use `SymPy` to find the formula for $d_n$. (Export the formula in LateX, do not worry if the formula looks ugly.)
 

<div markdown=1 class="Answers"> 

**Answer for Exercise 5.1**

Since we admit that for every $n\geq 1$ there exists three integers $a_n,b_n,c_n,d_n$ such that
$$
(\sqrt{2}+\sqrt{3})^n=a_n+b_n\sqrt{2}+c_n\sqrt{3}+d_n\sqrt{6}.
$$
Therefore we have the computation
\begin{align*}
(\sqrt{2}+\sqrt{3})^{n+1}&=a_{n+1}+b_{n+1}\sqrt{2}+c_{n+1}\sqrt{3}+d_{n+1}\sqrt{6}\\
&=(\sqrt{2}+\sqrt{3})(a_n+b_n\sqrt{2}+c_n\sqrt{3}+d_n\sqrt{6})\\
&=(2b_{n}+3c_{n})+(a_{n}+3d_{n})\sqrt{2}+(a_{n}+2d_{n})\sqrt{3}+(b_{n}+c_{n})\sqrt{6}
\end{align*}
Then we can have the relation that
\begin{cases}
a_{n+1}&=2b_n+3c_n,\\
b_{n+1}&=a_n+3d_n,\\
c_{n+1}&=a_n+2d_n,\\
d_{n+1}&=b_n+c_n.
\end{cases}
Therefore we can get the matrix $A$ as
$$
A=\begin{pmatrix}
0&2&3&0\\
1&0&0&3\\
1&0&0&2\\
0&1&1&0
\end{pmatrix}
$$
Also we have the initial value $a_1=0,b_1=1,c_1=1,d_1=0$.

In [5]:
# Question 2
A = Matrix([[0, 2, 3, 0], [1, 0, 0, 3], [1, 0, 0, 2], [0, 1, 1, 0]])
var('n', integer=True)
Power = A**(n - 1)

dn = simplify(Power[3, 1] + Power[3, 2])
display(simplify(dn))
print('-------------------------')
print('The latex formula of dn is ', latex(simplify(dn)))

(-5*(-sqrt(5 - 2*sqrt(6)))**n*sqrt(12*sqrt(6) + 30)/24 + (-sqrt(5 - 2*sqrt(6)))**n*sqrt(2*sqrt(6) + 5)/2 + sqrt(30 - 12*sqrt(6))*((-sqrt(2*sqrt(6) + 5))**n - (5 - 2*sqrt(6))**(n/2) + (2*sqrt(6) + 5)**(n/2))/24)/sqrt(5 - 2*sqrt(6))

-------------------------
The latex formula of dn is  \frac{- \frac{5 \left(- \sqrt{5 - 2 \sqrt{6}}\right)^{n} \sqrt{12 \sqrt{6} + 30}}{24} + \frac{\left(- \sqrt{5 - 2 \sqrt{6}}\right)^{n} \sqrt{2 \sqrt{6} + 5}}{2} + \frac{\sqrt{30 - 12 \sqrt{6}} \left(\left(- \sqrt{2 \sqrt{6} + 5}\right)^{n} - \left(5 - 2 \sqrt{6}\right)^{\frac{n}{2}} + \left(2 \sqrt{6} + 5\right)^{\frac{n}{2}}\right)}{24}}{\sqrt{5 - 2 \sqrt{6}}}


<div markdown=1 class="Answers"> 

**Answer for Exercise 5.2**

From above we have that
$$
\begin{pmatrix}
a_{n+1}\\
b_{n+1}\\
c_{n+1}\\
d_{n+1}
\end{pmatrix} = 
\begin{pmatrix}
0&2&3&0\\
1&0&0&3\\
1&0&0&2\\
0&1&1&0
\end{pmatrix} \times
\begin{pmatrix}
a_n\\
b_n\\
c_n\\
d_n
\end{pmatrix}
$$
which can be written as
$$
\begin{pmatrix}
a_{n+1}\\
b_{n+1}\\
c_{n+1}\\
d_{n+1}
\end{pmatrix} = 
\begin{pmatrix}
0&2&3&0\\
1&0&0&3\\
1&0&0&2\\
0&1&1&0
\end{pmatrix}^{n-1} \times
\begin{pmatrix}
a_1\\
b_1\\
c_1\\
d_1
\end{pmatrix} = 
\begin{pmatrix}
0&2&3&0\\
1&0&0&3\\
1&0&0&2\\
0&1&1&0
\end{pmatrix}^{n-1} \times
\begin{pmatrix}
0\\
1\\
1\\
0
\end{pmatrix}
$$
Then we can get the formula for $d_n$ as
$$
d_{n}=A_{4,2}^{n-1}+A_{4,3}^{n-1}
$$
We export the result as LateX code we got
$$
d_{n}= \frac{- \frac{5 \left(- \sqrt{5 - 2 \sqrt{6}}\right)^{n} \sqrt{12 \sqrt{6} + 30}}{24} + \frac{\left(- \sqrt{5 - 2 \sqrt{6}}\right)^{n} \sqrt{2 \sqrt{6} + 5}}{2} + \frac{\sqrt{30 - 12 \sqrt{6}} \left(\left(- \sqrt{2 \sqrt{6} + 5}\right)^{n} - \left(5 - 2 \sqrt{6}\right)^{\frac{n}{2}} + \left(2 \sqrt{6} + 5\right)^{\frac{n}{2}}\right)}{24}}{\sqrt{5 - 2 \sqrt{6}}}
$$
